cpp/daal/src/algorithms/linear_model/linear_model_predict_dense_default_batch_impl.i

```diff
@@ -108,60 +163,48 @@ services::Status PredictKernel<algorithmFPType, defaultDense, cpu>::compute(cons
     DAAL_CHECK_BLOCK_STATUS(betaRows)
     const algorithmFPType * beta = betaRows.get();
 
-    size_t numRowsInBlock = _numRowsInBlock;
-
-    if (numRowsInBlock < 1)
-    {
-        numRowsInBlock = 1;
-    }
-
     /* Calculate number of blocks of rows including tail block */
-    size_t numBlocks = numVectors / numRowsInBlock;
-    if (numBlocks * numRowsInBlock < numVectors)
-    {
-        numBlocks++;
-    }
+    size_t numBlocks = numVectors / blockSizeRows + !!(numVectors % blockSizeRows);
 
     SafeStatus safeStat;
-    /* Loop over input data blocks */
-    daal::threader_for(numBlocks, numBlocks, [=, &safeStat](int iBlock) {
-        size_t startRow = iBlock * numRowsInBlock;
-        size_t endRow   = startRow + numRowsInBlock;
-        if (endRow > numVectors)
-        {
-            endRow = numVectors;
-        }
-
-        DAAL_INT numRows = endRow - startRow;
+    TlsMem<algorithmFPType, cpu> tlsData(blockSizeRows * blockSizeColumns);
+    daal::threader_for(numBlocks, numBlocks, [&](int iBlock) {
+        size_t startRow       = iBlock * blockSizeRows;
+        size_t numRowsInBlock = (iBlock == numBlocks - 1) ? numVectors - startRow : blockSizeRows;
 
         DAAL_INT numFeatures = dataTable->getNumberOfColumns();
         DAAL_INT nAllBetas   = betaTable->getNumberOfColumns();
 
         Status s;
-        /* Retrieve data blocks associated with input and resulting tables */
-        ReadRows<algorithmFPType, cpu> dataRows(dataTable, startRow, endRow - startRow);
-        s = dataRows.status();
-        if (!s)
-        {
-            safeStat |= s;
-            return;
-        }
-        const algorithmFPType * dataBlock = dataRows.get();
 
-        WriteOnlyRows<algorithmFPType, cpu> responseRows(r, startRow, endRow - startRow);
-        s = responseRows.status();
-        if (!s)
-        {
-            safeStat |= s;
-            return;
-        }
+        WriteOnlyRows<algorithmFPType, cpu> responseRows(r, startRow, numRowsInBlock);
+        DAAL_CHECK_BLOCK_STATUS_THR(responseRows);
         algorithmFPType * responseBlock = responseRows.get();
         DAAL_INT * pnumResponses        = (DAAL_INT *)&numResponses;
+        DAAL_INT * pnumRowsInBlock      = (DAAL_INT *)&numRowsInBlock;
 
-        /* Calculate predictions */
-        computeBlockOfResponses(&numFeatures, &numRows, dataBlock, &nAllBetas, beta, pnumResponses, responseBlock, model->getInterceptFlag());
+        if (dataTable->getDataLayout() & NumericTableIface::soa)
+        {
+            /* Calculate predictions */
+            s = computeBlockOfResponsesSOA(startRow, &numFeatures, pnumRowsInBlock, dataTable, &nAllBetas, beta, pnumResponses, responseBlock,
+                                           model->getInterceptFlag(), tlsData.local());
```

Comment:
you shouldn't always cause allocation. pass the pointer to the tls, allocate only where necessary.

after the allocation, it is necessary to check that the memory is valid.